### ISSUE TITLE
Proposal: add default Region and Walk-area settings to the Room

### ISSUE THREAD

**Issue by ivan-mogilko** [adventuregamestudio/ags#2129]
This idea came to me when thinking about some setting that could be applied either globally for the whole game, or per room, or per walkable area.

AGS currently has certain properties applied to Regions and Walkable areas. I don't mention hotspots and walk-behinds, because hotspots are there only for interaction, and don't affect anything else, and walk-behinds are simply bg cut-outs. But Regions and Walkables affect Characters and Objects standing on them, with properties such as: light level, tint, scaling.

There may be situations when one would like to have whole room applying same effects to characters, or whole room except few places. If there are many areas, it may become a chore to assign identical settings everywhere. But things will become more complicated with continuous scaling property, for example, because if you have a lot of areas you would need to calculate bottom and top scaling levels correctly. And this becomes even more annoying if these areas have complex shapes.

So, what I'd like to propose for discussion is this idea:
* have the full copy of the Region and Walkable area properties inside the Room properties; these are considered "default";
* Regions and Walkable areas would only apply individual settings if they are enabled, otherwise Room's defaults are used.

This would refer to both in Editor and Script API properties.

Note: at first I've been thinking about using area 0 for this, but technically it's also a valid area, which may apply its settings when character stands on it. At least the walkable area 0 does (I don't remember about region 0).

**Comment by ericoporto** [comment]
So currently when a character is outside the room area and I want to slide the character in I first scale them to whatever is needed to match the scale of the first pixel they will match once their x,y position is the same as the scaling for what is set in the walkable area. After this change, does the set of default applies continuously to the character outside the walkable area, or only once and I can override by re-setting it?

**Comment by ivan-mogilko** [comment]
From the engine's perspective currently there's no such thing as "no area". You may assign scaling to area 0, or color and light to region 0, despite them being called "eraser" in the editor, and these parameters will actually work. You may test area 0 scaling by moving character in game using eAnywhere flag.

So, it achieve what you want you could assign values to the area 0 already (and in script too).

What I'm proposing is a default that works for all areas, so long as they don't have their own settings.

Of course this raises a concern about convenience, because right now if area has scaling 100, that would mean an explicit setting already. If the settings would be inherited from the parent (Room), and optionally overridden in areas, then there would have to be an explicit "enabling" property on each area. This is something I am not certain about.

**Comment by ivan-mogilko** [comment]
Actually... maybe we should move to factor multiplication instead of override in ags4.

Similar to how audio volume is a multiplication of factors: system volume * clip volume, or * object volume * animation volume
The scaling could be: room scaling * area scaling * character scaling.

Right now, if you want to give particular character a scaling in addition to the area scaling, you would have to make it manual, and calculate in script. With the system above it will be possible to apply one effect to another in chain.

**Comment by ericoporto** [comment]
Just to explain the issue of outside of room

![image](https://github.com/adventuregamestudio/ags/assets/2244442/4bd6ffc5-8ba3-4a2d-98fe-b94b437e7d45)

Usually when moving a character from outside of the room inside the room, the character will be big (or just at different scaling) until the middle of the feet hit the first walkable area pixel inside the room. Ideally, would be nice if the last pixel at right and first pixel at left were assumed to be the same for things out of room borders. Right now, the solution is to pre-set the character scaling manually, and then move it into the room - it will only start to active scale after being scaled once manually once it hits the pixel of the walkable area in room borders.

**Comment by ivan-mogilko** [comment]
> Ideally, would be nice if the last pixel at right and first pixel at left were assumed to be the same for things out of room borders.

Perhaps not only one pixel, the character or object may have a larger sprite, so may have to stretch out far.

But I think this is a kind of a separate problem. Global room settings won't necessarily solve this, as there may be multiple characters entering from different directions and walkable areas.

**Comment by ivan-mogilko** [comment]
>  Ideally, would be nice if the last pixel at right and first pixel at left were assumed to be the same for things out of room borders.

Oh, right, so imagine that walkable areas stretch infinitely behind the room borders.

Yes, this may be a good idea, but it's best to open a separate ticket for this.

**Comment by AlanDrake** [comment]
> imagine that walkable areas stretch infinitely behind the room borders.

I like this idea too. It's simple and fixes all common annoyances.

**Comment by morganwillcock** [comment]
Does that also imply that the pathfinding is aware of and takes into account the bleed area?
What happens if the shortest distance is through the bleed area?

**Comment by ericoporto** [comment]
No, not walkable, my worry is currently if one wants to make a character enter the room from outside of it, currently you set the scaling, and tween them in, and then make the character walk as soon as they are in a walkable area, but if there's some "default" scaling and it's applied to characters that are not in any walkable area (outside), this won't work anymore.

**Comment by ivan-mogilko** [comment]
Ermmm I am bit confused right now, I thought this was a feature suggestion about continuing area's effect beyond the room borders. Not sure anymore if I imagined this.

But regarding your concern:

1. If there's a "ignore area scaling" flag, and you are setting it per character, then this overrides anything anyway.
2. About the "default" scaling, at first I thought an override mechanic, where area always overrides room, and character always overrides area (if a flag is set). But, as I mentioned in one of the comments above, I am now leaning to multiplicative factors, but still with optionally letting to let individual characters and objects ignore the room's effect. I would like to think more about how this may be configured, but in any case, the ability to have completely overriding scaling should be preserved.
